<a href="https://colab.research.google.com/github/azamatarmanuly99/docs/blob/main/notebooks/12.01-RosettaAntibody-Framework-and-SimpleMetrics.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<!--NOTEBOOK_HEADER-->
*This notebook contains material from [PyRosetta](https://RosettaCommons.github.io/PyRosetta.notebooks);
content is available [on Github](https://github.com/RosettaCommons/PyRosetta.notebooks.git).*

<!--NAVIGATION-->
< [Working With Antibodies](http://nbviewer.jupyter.org/github/RosettaCommons/PyRosetta.notebooks/blob/master/notebooks/12.00-Working-With-Antibodies.ipynb) | [Contents](toc.ipynb) | [Index](index.ipynb) | [RosettaAntibodyDesign](http://nbviewer.jupyter.org/github/RosettaCommons/PyRosetta.notebooks/blob/master/notebooks/12.02-RosettaAntibodyDesign-RAbD.ipynb) ><p><a href="https://colab.research.google.com/github/RosettaCommons/PyRosetta.notebooks/blob/master/notebooks/12.01-RosettaAntibody-Framework-and-SimpleMetrics.ipynb"><img align="left" src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open in Colab" title="Open in Google Colaboratory"></a>

# RosettaAntibody Framework
Keywords: CDRResidueSelector

## Overview
In this workshop we will learn how to use the RosettaAntibody framework.  The full RosettaAntibody (modeling) code is not available in PyRosetta, unfortunately - as it is based around an application. To use that, you will have to use either the ROSIE server, or the Rosetta application.

For a full overview of the RosettaAntibody modeling application, see this paper:
https://www.ncbi.nlm.nih.gov/pubmed/28125104

Snugdock, and H3 modeling component of RosettaAntibody are available here as movers.

In [1]:
!pip install pyrosettacolabsetup
import pyrosettacolabsetup; pyrosettacolabsetup.install_pyrosetta()
import pyrosetta; pyrosetta.init()


Mounted at /content/google_drive
Looking for compatible PyRosetta wheel file at google-drive/PyRosetta/colab.bin//wheels...
Found compatible wheel: /content/google_drive/MyDrive/PyRosetta/colab.bin/wheels//content/google_drive/MyDrive/PyRosetta/colab.bin/wheels/pyrosetta-2024.1+release.00b79147e63-cp310-cp310-linux_x86_64.whl


PyRosetta-4 2023 [Rosetta PyRosetta4.MinSizeRel.python310.ubuntu 2024.01+release.00b79147e63be743438188f93a3f069ca75106d6 2023-12-25T16:35:48] retrieved from: http://www.pyrosetta.org
(C) Copyright Rosetta Commons Member Institutions. Created in JHU by Sergey Lyskov and PyRosetta Team.
core.init: Checking for fconfig files in pwd and ./rosetta/flags
core.init: Rosetta version: PyRosetta4.MinSizeRel.python310.ubuntu r366 2024.01+release.00b79147e63 00b79147e63be743438188f93a3f069ca75106d6 http://www.pyrosetta.org 2023-12-25T16:35:48
core.init: command: PyRosetta -ex1 -ex2aro -database /usr/local/lib/python3.10/dist-packages/pyrosetta/database
basic.random.init_ra

**Make sure you are in the directory with the pdb files:**

`cd google_drive/MyDrive/student-notebooks/`

## Imports

Lets import the antibody namespace so we can start using it.  Take a look at the different modules that are a part of the antibody module.

Note that we can also do `from rosetta.protocols.antibody import *` in order to make accessing the enums much easier.  For the purpose of this workshop, we will use `antibody` to traverse the contents.  This makes it easier for you to use tab completion for exploration.

In [2]:
#Python
from pyrosetta import *
from pyrosetta.rosetta import *
from pyrosetta.teaching import *

#Core Includes
from rosetta.core.select import residue_selector as selections

from rosetta.protocols import antibody


<ipython-input-2-03bff5fc4534>:7: UserWarning: Import of 'rosetta' as a top-level module is deprecated and may be removed in 2018, import via 'pyrosetta.rosetta'.
  from rosetta.core.select import residue_selector as selections


## Intitlialization

Here, we will initialize a typical run of Rosetta. We could use the `-input_ab_scheme` option with `AHo_Scheme`, but we will learn to instead pass this to our main antibody framework code.

In [3]:
init('-use_input_sc -ignore_unrecognized_res \
     -ignore_zero_occupancy false -load_PDB_components false -no_fconfig')

PyRosetta-4 2023 [Rosetta PyRosetta4.MinSizeRel.python310.ubuntu 2024.01+release.00b79147e63be743438188f93a3f069ca75106d6 2023-12-25T16:35:48] retrieved from: http://www.pyrosetta.org
(C) Copyright Rosetta Commons Member Institutions. Created in JHU by Sergey Lyskov and PyRosetta Team.
core.init: Rosetta version: PyRosetta4.MinSizeRel.python310.ubuntu r366 2024.01+release.00b79147e63 00b79147e63be743438188f93a3f069ca75106d6 http://www.pyrosetta.org 2023-12-25T16:35:48
core.init: command: PyRosetta -use_input_sc -ignore_unrecognized_res -ignore_zero_occupancy false -load_PDB_components false -no_fconfig -database /usr/local/lib/python3.10/dist-packages/pyrosetta/database
basic.random.init_random_generator: 'RNG device' seed mode, using '/dev/urandom', seed=796582631 seed_offset=0 real_seed=796582631
basic.random.init_random_generator: RandomGenerator:init: Normal mode, seed=796582631 RG_type=mt19937


## Import and copy pose

Let's load an antibody - this this the same antibody we used to learn packing and design. :)

In [4]:
#Import a pose
pose = pose_from_pdb("2r0l_1_1.pdb")
original_pose = pose.clone()

core.chemical.GlobalResidueTypeSet: Finished initializing fa_standard residue type set.  Created 985 residue types
core.chemical.GlobalResidueTypeSet: Total time to initialize 1.08531 seconds.
core.import_pose.import_pose: File '2r0l_1_1.pdb' automatically determined to be of type PDB
core.conformation.Conformation: [ WARNING ] missing heavyatom:  OXT on residue ARG:CtermProteinFull 108
core.conformation.Conformation: [ WARNING ] missing heavyatom:  OXT on residue SER:CtermProteinFull 225
core.conformation.Conformation: [ WARNING ] missing heavyatom:  OXT on residue ARG:CtermProteinFull 464
core.conformation.Conformation: Found disulfide between residues 23 88
core.conformation.Conformation: Found disulfide between residues 130 204
core.conformation.Conformation: Found disulfide between residues 250 266
core.conformation.Conformation: Found disulfide between residues 258 328
core.conformation.Conformation: Found disulfide between residues 353 422
core.conformation.Conformation: Found d

## AntibodyInfo

The main tool that we will use is the `AntibodyInfo` object.  This allows you to get a TON of information about the antibody to use in various custom protocols.  

Note that this antibody has already been renumbered using the PyIgClassify server.

Since we are not defining the numbering scheme and cdr definition during init, we will need to pass an Enum to the AntibodyInfo object.

In [5]:
ab_info = antibody.AntibodyInfo(pose, antibody.AHO_Scheme, antibody.North)

basic.io.database: Database file opened: sampling/antibodies/cluster_center_dihedrals.txt
protocols.antibody.AntibodyNumberingParser: Antibody numbering scheme definitions read successfully
protocols.antibody.AntibodyNumberingParser: Antibody CDR definition read successfully
antibody.AntibodyInfo: Successfully finished the CDR definition
antibody.AntibodyInfo: AC Detecting Regular CDR H3 Stem Type
antibody.AntibodyInfo: ARFWWRSFDYW
antibody.AntibodyInfo: AC Finished Detecting Regular CDR H3 Stem Type: KINKED
antibody.AntibodyInfo: AC Finished Detecting Regular CDR H3 Stem Type: Kink: 1 Extended: 0
antibody.AntibodyInfo: Setting up CDR Cluster for H1
protocols.antibody.cluster.CDRClusterMatcher: Length: 13 Omega: TTTTTTTTTTTTT
antibody.AntibodyInfo: Setting up CDR Cluster for H2
protocols.antibody.cluster.CDRClusterMatcher: Length: 10 Omega: TTTTTTTTTT
antibody.AntibodyInfo: Setting up CDR Cluster for H3
protocols.antibody.cluster.CDRClusterMatcher: Length: 10 Omega: TTTTTTTTTT
antibody

Lets take a look at what AntibodyInfo prints

In [6]:
print(ab_info)

////////////////////////////////////////////////////////////////////////////////
///                          Rosetta Antibody Info                           ///
///                                                                          ///
///             Antibody Type:  Regular Antibody
///             Light Chain Type:  unknown
/// Predict H3 Cterminus Base:  KINKED
///                                                                          
/// H1 info: 
///            length:  13
///          sequence:  AASGFTISNSGIH
///     north_cluster:  H1-13-1
///         loop_info:  LOOP start: 131  stop: 143  cut: 137  size: 13  skip rate: 0  extended?: False

/// H2 info: 
///            length:  10
///          sequence:  WIYPTGGATD
///     north_cluster:  H2-10-1
///         loop_info:  LOOP start: 158  stop: 167  cut: 163  size: 10  skip rate: 0  extended?: False

/// H3 info: 
///            length:  10
///          sequence:  ARFWWRSFDY
///     north_cluster:  H3-10-1
///         l

**Isn't that AWESOME!!**  I think so.  But I wrote a lot of that code!  

Anyway, as you can see you can get a pretty fair bit of information out of the AntibodyInfo object.  In fact, most antibody-related code actually takes an AntibodyInfo object or constructs one from set numbering scheme, cdr definitions, and pose passed to it.  You will see this as we go.  

Note the north_cluster here.  This is useful in some modeling tasks, but becomes much more relevant during antibody design.  More information on what we mean by north_cluster can be found in this paper, if you want to read ahead a bit. https://www.ncbi.nlm.nih.gov/pubmed/21035459

## Basic AntibodyInfo Access
Now, lets use the AntibodyInfo class to get a bit of useful information out of our antibody.

In [7]:
print("h1", ab_info.get_CDR_start(antibody.h1, pose))
print("h2", ab_info.get_CDR_end(antibody.h2, pose))

h1 131
h2 167


Now lets use these enums a bit more.  They go in order from 1 to 8, with 7 and 8 being CDR4 loops - also known as H3 loops.  We won't worry about them just yet.  

In [8]:
for i in range(1, 7):
    print(i, ab_info.get_CDR_name(antibody.CDRNameEnum(i)))

for cdr in ['L1', 'l1', 'L2', 'l2', 'L3', 'H1', 'H2', 'H3']:
    print(cdr, str(ab_info.get_CDR_name_enum(cdr)))

print(str(antibody.h3))
print(int(antibody.h3))

1 H1
2 H2
3 H3
4 L1
5 L2
6 L3
L1 CDRNameEnum.l1
l1 CDRNameEnum.l1
L2 CDRNameEnum.l2
l2 CDRNameEnum.l2
L3 CDRNameEnum.l3
H1 CDRNameEnum.h1
H2 CDRNameEnum.h2
H3 CDRNameEnum.h3
CDRNameEnum.h3
3


Does this make enums a bit less confusing?  These are named integers.  The last function allows us to print either the actual cdr name enum or the integer from it.  The cool thing here is that we can loop through all of the CDRs just by using a range 1-6 and rosetta will understand it.  

Note that we convert the integer into a `CDRNameEnum` in the function.  If we are storing the cdr name enums as indexes to a dictionary or list, we don't need this.  That is simply for the C++ code to work properly.

### AntibodyEnumManager
So we have seen that some of this code we can do directly within AntibodyInfo itself.  Cool. But what if we need something more advanced?  Lets use the class that actually does all this conversion.


In [9]:
enum_manager = antibody.AntibodyEnumManager()
print(enum_manager.numbering_scheme_enum_to_string(antibody.AHO_Scheme))
print(enum_manager.cdr_definition_enum_to_string(antibody.North))
print(enum_manager.cdr_name_string_to_enum("H1"))
print(enum_manager.antibody_region_enum_to_string(antibody.framework_region))

AHO_Scheme
North
CDRNameEnum.h1
framework_region


Use the function, `get_region_or_residue` and `get_CDRNameEnum_of_residue` and the manager to traverse the antibody and get relevant regions of all residues in the pose

In [10]:
### BEGIN SOLUTION

for i in range(1, pose.size()+1):
    region = ab_info.get_region_of_residue(pose, i)
    if (region == antibody.cdr_region):
        print(i, enum_manager.cdr_name_enum_to_string(ab_info.get_CDRNameEnum_of_residue(pose, i)))
    else:
        print(i, enum_manager.antibody_region_enum_to_string(region))

### END SOLUTION

1 framework_region
2 framework_region
3 framework_region
4 framework_region
5 framework_region
6 framework_region
7 framework_region
8 framework_region
9 framework_region
10 framework_region
11 framework_region
12 framework_region
13 framework_region
14 framework_region
15 framework_region
16 framework_region
17 framework_region
18 framework_region
19 framework_region
20 framework_region
21 framework_region
22 framework_region
23 framework_region
24 L1
25 L1
26 L1
27 L1
28 L1
29 L1
30 L1
31 L1
32 L1
33 L1
34 L1
35 framework_region
36 framework_region
37 framework_region
38 framework_region
39 framework_region
40 framework_region
41 framework_region
42 framework_region
43 framework_region
44 framework_region
45 framework_region
46 framework_region
47 framework_region
48 framework_region
49 L2
50 L2
51 L2
52 L2
53 L2
54 L2
55 L2
56 L2
57 framework_region
58 framework_region
59 framework_region
60 framework_region
61 framework_region
62 framework_region
63 framework_region
64 framework_re

### CDR Clusters

Use either the PyRosetta docs on AntibodyInfo, or the interactive notebook to use AntibodyInfo to get the length and cluster of L1.

In [11]:
### BEGIN SOLUTION

print(ab_info.get_CDR_length(antibody.l1))
print(ab_info.get_CDR_cluster(antibody.l1).cluster())

### END SOLUTION

11
CDRClusterEnum.L1_11_1


The CDRCluster object has a lot of information about a particular cluster.  Lets use it to get the normalized distance in degrees of the L1 cluster.

In [12]:
L1_cluster = ab_info.get_CDR_cluster(antibody.l1)
print(L1_cluster.normalized_distance_in_degrees())

7.1372427840879435


Anything below 35 or 40 degrees is very close to the cluster center.  This is a structure with a very well-defined L1-11-1 loop - one of the most common L1 lengths and clusters.

### Numbering Scheme Translation
It may not seem like much, but numbering scheme translation is a very difficult thing to do without mistakes.   Rosetta now has this ability to make it much easier to understand antibody structural or sequence papers in a highly tested and fairly easy-to-use implementation.  Lets take a look.  We'll use `AntibodyInfo` and the `get_landmark_resnum()` function for this, but you could also use function `get_antibody_numbering_info()` that will give you all the conversions - though it is certainly a bit more tricky to use.

#### Conserved Inter-Domain Cysteine

The conserved cysteine residue forming the intradomain disulfide bridge always carries the label "23" as in the IMGT numbering scheme, while according to Kabat, it was labeled L23 in Vk and Vl, H22 in VH.  Let's find this residue in our antibody. https://www.bioc.uzh.ch/plueckthun/antibody/Numbering/FR1/index.html

In [13]:
rosetta_num = ab_info.get_landmark_resnum(pose, antibody.Kabat_Scheme, 'H', 22)

What is the chain and resnum in OUR Aho numbering scheme?  Is this a cysteine?  How about a disulfide?

In [14]:
### BEGIN SOLUTION
print(pose.pdb_info().pose2pdb(rosetta_num))
print(pose.residue(rosetta_num))
### END SOLUTION

23 H 
Residue 130: CYS:disulfide (CYS, C):
Base: CYS
 Properties: POLYMER PROTEIN CANONICAL_AA SC_ORBITALS METALBINDING DISULFIDE_BONDED ALPHA_AA L_AA
 Variant types: DISULFIDE
 Main-chain atoms:  N    CA   C  
 Backbone atoms:    N    CA   C    O    H    HA 
 Side-chain atoms:  CB   SG  1HB  2HB 
Atom Coordinates:
   N  : -13.918, -0.011, 40.022
   CA : -15.022, -0.943, 39.837
   C  : -16.073, -0.624, 40.895
   O  : -15.877, -0.945, 42.066
   CB : -14.515, -2.379, 40.021
   SG : -15.8, -3.608, 40.319
   H  : -13.6187, 0.217354, 40.9592
   HA : -15.4065, -0.826975, 38.8236
  1HB : -13.9648, -2.68746, 39.1317
  2HB : -13.8236, -2.41565, 40.8626
Mirrored relative to coordinates in ResidueType: FALSE



Ok.  Cool.  Lets do the same thing for the Cysteine that is connected to this residue.
In IMGT this is residue 104 on the heavy chain.  Lets do the same thing here.  Use tab completion for  `antibody.IMGT_Scheme` for the enum.  https://www.bioc.uzh.ch/plueckthun/antibody/Numbering/FR3a/index.html

In [15]:
### BEGIN SOLUTION
pre_cdr3_c = ab_info.get_landmark_resnum(pose, antibody.IMGT_Scheme, 'H', 104)
### END SOLUTION

Once again, what is the residue in our AHO-numbered antibody?  Is it a Cysteine?  Is it disulfide bonded?

In [16]:
### BEGIN SOLUTION
print(pose.pdb_info().pose2pdb(pre_cdr3_c))
print(pose.residue(pre_cdr3_c))

### END SOLUTION

106 H 
Residue 204: CYS:disulfide (CYS, C):
Base: CYS
 Properties: POLYMER PROTEIN CANONICAL_AA SC_ORBITALS METALBINDING DISULFIDE_BONDED ALPHA_AA L_AA
 Variant types: DISULFIDE
 Main-chain atoms:  N    CA   C  
 Backbone atoms:    N    CA   C    O    H    HA 
 Side-chain atoms:  CB   SG  1HB  2HB 
Atom Coordinates:
   N  : -14.312, -6.402, 36.316
   CA : -14.452, -6.929, 37.646
   C  : -15.678, -7.837, 37.662
   O  : -16.599, -7.672, 36.856
   CB : -14.501, -5.824, 38.705
   SG : -15.935, -4.767, 38.638
   H  : -14.9281, -5.66132, 36.0129
   HA : -13.5885, -7.5585, 37.8613
  1HB : -14.4721, -6.27099, 39.699
  2HB : -13.6222, -5.18697, 38.607
Mirrored relative to coordinates in ResidueType: FALSE



### Sequence
Lets expore the sequence of this antibody

In [17]:
ab_seq = ab_info.get_antibody_sequence()
print(ab_seq)

L1_seq = ab_info.get_CDR_sequence_with_stem(antibody.l1, pose)
print("L1", L1_seq)

for i in range(1, 7):
    cdr = antibody.CDRNameEnum(i)
    print(cdr, ab_info.get_CDR_sequence_with_stem(cdr, pose))

DIQMTQSPSSLSASVGDRVTITCRASQDVSTAVAWYQQKPGKAPKLLIYSASFLYSGVPSRFSGSGSGTDFTLTISSLQPEDFATYYCQQSYTTPPTFGQGTKVEIKREVQLVESGGGLVQPGGSLRLSCAASGFTISNSGIHWVRQAPGKGLEWVGWIYPTGGATDYADSVKGRFTISADTSKNTAYLQMNSLRAEDTAVYYCARFWWRSFDYWGQGTLVTVSS
L1 RASQDVSTAVA
CDRNameEnum.h1 AASGFTISNSGIH
CDRNameEnum.h2 WIYPTGGATD
CDRNameEnum.h3 ARFWWRSFDY
CDRNameEnum.l1 RASQDVSTAVA
CDRNameEnum.l2 YSASFLYS
CDRNameEnum.l3 QQSYTTPPT


### Other AntibodyInfo functions
Use tab completion to find other useful functions.  This includes movemap, loops, and fold-tree creation for specific tasks.  With ResidueSelectors, this functionality is not quite as useful, but you should know that it is here.

### AntibodyInfo Deprecated Functions
All functions are fair-game, except these: `get_TaskFactory_AllCDRs` and `get_TaskFactory_OneCDR`  - This will be removed from AntibodyInfo as it is extremely specific to a particular antibody modeling task.

## Antibody Util and SimpleMetrics
Util functions in Rosetta are stored in the `util.hh` file in each directory that has one.  Within PyRosetta, when you import the namespace, these come with.  There are many that you should be aware of to make modeling and design tasks easier for custom protocols.

We will go through some examples here.

### Function: get_cdr_loops()
The get_cdr_loops function takes a vector1 bool of CDRs.  Use the Enums to set H3 and L3 to true.
Vector1 bool starts as all negative.

In [18]:
h3_l3 = rosetta.utility.vector1_bool(6)
print(h3_l3)

h3_l3[antibody.h3] = True
h3_l3[antibody.l3] = True

#Here, we get cdr loops, and set the stem size to 2,
# so we include 2 residues on either side of the CDR loop (called the stem), to help us in modeling.
h3_l3_loops = antibody.get_cdr_loops(ab_info, pose, h3_l3, 2)
print(h3_l3_loops)

vector1_bool[0, 0, 0, 0, 0, 0]
LOOP  begin  end  cut  skip_rate  extended
LOOP start: 203  stop: 216  cut: 210  size: 14  skip rate: 0  extended?: False

LOOP start: 87  stop: 99  cut: 93  size: 13  skip rate: 0  extended?: False




### Function: select_epitope_residues()
We could use the NeighborhoodResidueSelector as you have used in the passed to get neighbors.  Instead, lets use a general function to get all the epitope residues within an 8 Angstrum distance of the paratope.

In [19]:
epi_residues = antibody.select_epitope_residues(ab_info, pose, 8)
total=0
for i in range(1, len(epi_residues)+1):
    if epi_residues[i]:
        print(i)
        total+=1
print("Total Epitope Residues:", total)

267
270
271
272
273
299
300
301
302
303
304
305
307
308
309
310
313
396
397
398
454
458
Total Epitope Residues: 22


So that was cool.  But lets the wonderful `ReturnResidueSubsetSelector` to take this `ResidueSubset` of the epitope residues and store the data as a `ResidueSelector`!

In [20]:
epi_res_selector = selections.ReturnResidueSubsetSelector(epi_residues)

Now what?  Lets use some SimpleMetrics using the selector to calculate something about these epitope residues.

### SasaMetric, TotalEnergyMetric, SelectedResiduesPyMOLMetric

In [21]:
import rosetta.core.simple_metrics.metrics as sm
sasa_metric = sm.SasaMetric(epi_res_selector)
print("\nSASA", sasa_metric.calculate(pose))

total_metric = sm.TotalEnergyMetric(epi_res_selector)
print("\nTOTAL RESIDUE ENERGY", total_metric.calculate(pose))

#Lets use a useful metric to select these residues in pymol
pymol_metric = sm.SelectedResiduesPyMOLMetric(epi_res_selector)
print("\nSELECTION", pymol_metric.calculate(pose))


SASA 531.9639835627297
core.scoring.ScoreFunctionFactory: SCOREFUNCTION: ref2015
core.scoring.etable: Starting energy table calculation
core.scoring.etable: smooth_etable: changing atr/rep split to bottom of energy well
core.scoring.etable: smooth_etable: spline smoothing lj etables (maxdis = 6)
core.scoring.etable: smooth_etable: spline smoothing solvation etables (max_dis = 6)
core.scoring.etable: Finished calculating energy tables.
basic.io.database: Database file opened: scoring/score_functions/hbonds/ref2015_params/HBPoly1D.csv
basic.io.database: Database file opened: scoring/score_functions/hbonds/ref2015_params/HBFadeIntervals.csv
basic.io.database: Database file opened: scoring/score_functions/hbonds/ref2015_params/HBEval.csv
basic.io.database: Database file opened: scoring/score_functions/hbonds/ref2015_params/DonStrength.csv
basic.io.database: Database file opened: scoring/score_functions/hbonds/ref2015_params/AccStrength.csv
basic.io.database: Database file opened: scoring/

Now lets see which of these residues are most buried in the interface and the residues which have the lowest energy. Note that this is not ddG - we would need to separate the chains for this.  We can use the `protocols.toolbox.rigid_body.translate` function to do that.

Use the pymol selection (copy from select...) and lets take a look at them in PyMol.  Then run the code below.

### PerResidueSasaMetric

In [22]:
import rosetta.core.simple_metrics.per_residue_metrics as residue_sm
import operator

res_sasa_metric = residue_sm.PerResidueSasaMetric()
res_sasa_metric.set_residue_selector(epi_res_selector)
per_res_sasa = res_sasa_metric.calculate(pose)
#print(per_res_sasa)

#Convert the Map to a Dictionary, which are essentially the same thing.
for ele in sorted(per_res_sasa.items(), key=operator.itemgetter(1), reverse=False):
    print(ele)

(300, 0.0)
(303, 0.0)
(305, 0.0)
(304, 0.4468042885105504)
(267, 1.024686682355324)
(398, 1.8244508447514138)
(302, 4.098746729421277)
(396, 4.098746729421277)
(271, 5.380780270728442)
(307, 5.504698647620041)
(301, 7.689850871116937)
(313, 8.068377879289471)
(270, 8.322490061360945)
(458, 14.530261630816586)
(310, 17.873691916125896)
(308, 39.92505047193878)
(454, 46.22431607929343)
(397, 54.69632267990853)
(299, 60.31999848338232)
(309, 75.13475653929288)
(272, 76.34620896564937)
(273, 100.45374379174626)


Cool. So the most buried residues at the interface are 300, 303, 305.  Convert those to the PDB chain/num using PDBInfo and take a look at them in PyMOL.

### PerResidueEnergyMetric

In [23]:
res_energy_metric = residue_sm.PerResidueEnergyMetric()
res_energy_metric.set_residue_selector(epi_res_selector)

per_res_energy = res_sasa_metric.calculate(pose)
#print(per_res_sasa)

#Convert the Map to a Dictionary, which are essentially the same thing.
for ele in sorted(per_res_energy.items(), key=operator.itemgetter(1), reverse=False):
    print(ele[0], pose.pdb_info().pose2pdb(ele[0]), ele[1])

300 75 A  0.0
303 78 A  0.0
305 80 A  0.0
304 79 A  0.4468042885105504
267 42 A  1.024686682355324
398 173 A  1.8244508447514138
302 77 A  4.098746729421277
396 171 A  4.098746729421277
271 46 A  5.380780270728442
307 82 A  5.504698647620041
301 76 A  7.689850871116937
313 88 A  8.068377879289471
270 45 A  8.322490061360945
458 233 A  14.530261630816586
310 85 A  17.873691916125896
308 83 A  39.92505047193878
454 229 A  46.22431607929343
397 172 A  54.69632267990853
299 74 A  60.31999848338232
309 84 A  75.13475653929288
272 47 A  76.34620896564937
273 48 A  100.45374379174626


Wow!  Why is 48A so high in energy!?  This may be due to the fact that we are working with a crystal structure that has not been pre-relaxed using the pareto-optimal protocol.  Be sure when using PDBs from the data bank for production runs to do this, outputting about 10 models and selecting the lowest energy residue.  Or, you could use density to relax within the crystal denstiy.  Either works well.

https://journals.plos.org/plosone/article?id=10.1371/journal.pone.0059004

## Other Useful Antibody Tools

### CDRResidueSelector

In [24]:
from rosetta.protocols.antibody.residue_selector import *

cdr_selector = CDRResidueSelector(ab_info)
cdr_selector.set_cdrs(h3_l3)
sele = cdr_selector.apply(pose)
for i in range(1, len(sele)):
    if sele[i]:
        print(i, pose.pdb_info().pose2pdb(i))

89 107 L 
90 108 L 
91 109 L 
92 110 L 
93 111 L 
94 135 L 
95 136 L 
96 137 L 
97 138 L 
205 107 H 
206 108 H 
207 109 H 
208 110 H 
209 111 H 
210 134 H 
211 135 H 
212 136 H 
213 137 H 
214 138 H 


### AntibodyRegionSelector
We can use the AntibodyRegionSelector to select a specific region:
`antigen_region`, `framework_region`, and `cdr_region`

In [25]:
region_selector = AntibodyRegionSelector(ab_info)
region_selector.set_region(antibody.antigen_region)
sele = region_selector.apply(pose)

for i in range(1, len(sele)):
    if sele[i]:
        print(i, pose.pdb_info().pose2pdb(i))


226 1 A 
227 2 A 
228 3 A 
229 4 A 
230 5 A 
231 6 A 
232 7 A 
233 8 A 
234 9 A 
235 10 A 
236 11 A 
237 12 A 
238 13 A 
239 14 A 
240 15 A 
241 16 A 
242 17 A 
243 18 A 
244 19 A 
245 20 A 
246 21 A 
247 22 A 
248 23 A 
249 24 A 
250 25 A 
251 26 A 
252 27 A 
253 28 A 
254 29 A 
255 30 A 
256 31 A 
257 32 A 
258 33 A 
259 34 A 
260 35 A 
261 36 A 
262 37 A 
263 38 A 
264 39 A 
265 40 A 
266 41 A 
267 42 A 
268 43 A 
269 44 A 
270 45 A 
271 46 A 
272 47 A 
273 48 A 
274 49 A 
275 50 A 
276 51 A 
277 52 A 
278 53 A 
279 54 A 
280 55 A 
281 56 A 
282 57 A 
283 58 A 
284 59 A 
285 60 A 
286 61 A 
287 62 A 
288 63 A 
289 64 A 
290 65 A 
291 66 A 
292 67 A 
293 68 A 
294 69 A 
295 70 A 
296 71 A 
297 72 A 
298 73 A 
299 74 A 
300 75 A 
301 76 A 
302 77 A 
303 78 A 
304 79 A 
305 80 A 
306 81 A 
307 82 A 
308 83 A 
309 84 A 
310 85 A 
311 86 A 
312 87 A 
313 88 A 
314 89 A 
315 90 A 
316 91 A 
317 92 A 
318 93 A 
319 94 A 
320 95 A 
321 96 A 
322 97 A 
323 98 A 
324 99 A 
325 100 A 
326 101 

### Other
-  **TaskOperations** - Antibody specific Task Operations will be covered in the next workshop
-  **SnugDock** - Snugdock is available in the `rosetta.protocols.antibody.snugdock` namespace.  Both the full protocol, `SnugDockProtocol` and the mover, `Snugdock` are available and easy to setup through code - but their run time is extremely long.
-  **AntibodyModelerProtocol** - this is the `Antibody_H3` app. Personally, I would use the Rosetta C++ application for this with specific options specified in the docs, however you can call this in PyRosetta.
-  **AntibodyCDRGrafter** This is the main grafting class used for RosettaAntibody and RosettaAntibodyDesign. Is is in the `protocols.antibody` namespace. Documentation on this mover can be found here (XML or code-level interface is available): https://www.rosettacommons.org/docs/latest/scripting_documentation/RosettaScripts/Movers/movers_pages/antibodies/AntibodyCDRGrafter   

## References
Please site these papers when using any of RosettaAntibody.

- J. Adolf-Bryfogle, O Kalyuzhniy, M Kubitz, B. D. Weitzner, X Hu, Y Adachi, W R. Schief, R L. Dunbrack Jr.,
    - "Rosetta Antibody Design (RAbD): A General Framework for Computational Antibody Design", PLOS Computational Biology (2018)

- B. D. Weitzner*, J. R. Jeliazkov*, S. Lyskov*, N. M. Marze, D. Kuroda, R. Frick, J. Adolf-Bryfogle, N. Biswas, R. L. Dunbrack Jr., and J. J. Gray,
    - "Modeling and docking of antibody structures with Rosetta." Nature Protocols 12, 401–416 (2017)

- B. D. Weitzner, D. Kuroda, N. M. Marze, J. Xu & J. J. Gray,
    - "Blind prediction performance of RosettaAntibody 3.0: Grafting, relaxation, kinematic loop modeling, and full CDR optimization." Proteins 82(8), 1611–1623 (2014)

- A. Sivasubramanian,* A. Sircar,* S. Chaudhury & J. J. Gray,
    - "Toward high-resolution homology modeling of antibody Fv regions and application to antibody-antigen docking," Proteins 74(2), 497–514 (2009)

<!--NAVIGATION-->
< [Side Chain Conformations and Dunbrack Energies](http://nbviewer.jupyter.org/github/RosettaCommons/PyRosetta.notebooks/blob/master/notebooks/06.01-Side-Chain-Conformations-and-Dunbrack-Energies.ipynb) | [Contents](toc.ipynb) | [Index](index.ipynb) | [Protein Design with a Resfile and FastRelax](http://nbviewer.jupyter.org/github/RosettaCommons/PyRosetta.notebooks/blob/master/notebooks/06.03-Design-with-a-resfile-and-relax.ipynb) ><p><a href="https://colab.research.google.com/github/RosettaCommons/PyRosetta.notebooks/blob/master/notebooks/06.02-Packing-design-and-regional-relax.ipynb"><img align="left" src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open in Colab" title="Open in Google Colaboratory"></a>

<!--NAVIGATION-->
< [Working With Antibodies](http://nbviewer.jupyter.org/github/RosettaCommons/PyRosetta.notebooks/blob/master/notebooks/12.00-Working-With-Antibodies.ipynb) | [Contents](toc.ipynb) | [Index](index.ipynb) | [RosettaAntibodyDesign](http://nbviewer.jupyter.org/github/RosettaCommons/PyRosetta.notebooks/blob/master/notebooks/12.02-RosettaAntibodyDesign-RAbD.ipynb) ><p><a href="https://colab.research.google.com/github/RosettaCommons/PyRosetta.notebooks/blob/master/notebooks/12.01-RosettaAntibody-Framework-and-SimpleMetrics.ipynb"><img align="left" src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open in Colab" title="Open in Google Colaboratory"></a>